eip-0040.md

```diff
@@ -0,0 +1,69 @@
+4. efficiently query emission amount
+
+### EIP-40-Asset-ID
+
+Due to Ergo protocol design implications discussed earlier we can't have reproducible assets that share the same `tokenId`. To bypass this we introduce a synthetic token ID (let's call it EIP-40-Asset-ID). This helps to hide all the added complexity from the end user so managing the asset for them can be as convinient as manging EIP-4 asset.
+
+We define `EIP-40-Asset-ID` of an asset with `tokenId` as `EIP-40-Asset-ID(tokenId) = sha256(Box{boxId=tokenId}.ergoTreeBytes)`, where `Box{boxId=tokenId}` is the first input of the TX that mint the token `tokenId`, let's call it a "PermittingBox".
+
+Anchoring `EIP-40-Asset-ID` to a specific ErgoTree means that we can have something like "Minting Scripts", e.g. controll emission with a script. In the simplest case it can be `ProveDlog` - a proposition of some secret, in most complicated scenarios - a DAO script, which authorizes changing emission (e.g. minting) via voting.
+
+Querying proof of authorization can be done as follows `isAuthed(tokenId) := getBox(boxId=tokenId).isDefined && getUtxo(boxId=tokenId).isEmpty`
+
+In order to simplify the process of querying asset metadata we propose to store it in the PermittingBox. Then asset metadata can be queried as follows `metadata(EIP-40-Asset-ID(tokenId)) = parseEip40Metadata(getBox(boxId=tokenId))`, where `parseEip40Metadata` is a method that extracts all the EIP-4 fields from the given box, plus current amount of emission.
+
+Emission amount as well as other metadata members is written by permitting entity
+
+Let's define extended registers mapping from EIP-4:
+
+| Register        | Description                                     | Example                      |Encoded                      |
+| --------------- |:-----------------------------------------------:| ----------------------------:|----------------------------:|
+| R2              | Token id and amount pair                        | [("7d...09", 100000)]        |                             |
+| R4              | Token verbose name (UTF-8 representation)       | "USD"                        | "0e03555344"                |
+| R5              | Token description (UTF-8 representation)        | "Nothing backed USD token"   | "0e184e6f7468696e67206261636b65642055534420746f6b656e"  |
+| R6              | Number of decimals                              | "2"                          | "0e0132"                         |
+| R7              | Current emission                              | "1000000"                          | "0e0131303030303030"                         |
+
+\* An extension of EIP-4 for NFTs is not relevant here, so we can utilize R7.
+
+Asset metadata is controlled by an entity that permits minting (some party, in the case of simple `ProveDlog` or a script in more complex cases). Then, emission amount can be trustlessly checked by executing a query:
+```ts
+let permittingBoxes     = getBoxes(ergoTreeHash=EIP40AssetID)
```

Comment:
Could you please clarify this `getBoxes` id implemented, i.e. how the permittingBoxes are collected by EIP40AssetID?

Reply:
I assumed implementing similar method in explorer API. Just like this one we already have for ErgoTreeTemplate https://api.ergoplatform.com/api/v1/docs/#operation/getApiV1BoxesByergotreetemplatehashP1

Reply:
In the case of ergotree template you compare the template bytes with ergoTree bytes stored in DB. But here there is EIP40AssetID which is sha256 hash of ergotree bytes, so how will you search for the bytes in DB having just the hash?

Reply:
@aslesarenko in the provided method we compare hashes of template bytes (using index `sha256(box.ergoTree.template)`). Similar method for ErgoTree hash can be easily implemented.